kernel/vars.ml

```diff
@@ -228,6 +248,15 @@ let substn_vars p vars c =
 
 let subst_vars subst c = substn_vars 1 subst c
 
+let smash_rel_context sign =
```

Comment:
This is dead code in master, and to add insult to injury it's consciously quadratic. Can't you simply remove it?

Reply:
The problem is that I need it in #14798. It is also quadratic in the number of lets, so a priori not a lot. I thought about a non quadratic version but what I found was doing more work when no lets was there, which seems to be the general case to favor, so I renounced.

Reply:
What I'm surprised about is that you don't need to return the final substitution containing all the let bodies. This should be used for anything that lives under the context. Also I didn't find any use of that function in the code diff of #14798. Are you sure you need it?

Reply:
[Here](https://github.com/coq/coq/blob/7d853f77fadb875fab6debaea1991913e261868d/pretyping/evarsolve.ml#L604)

Reply:
Note that `smash_rel_context` is used in Coq-Equations:
- once casting the Coq version to `econstr` (e.g. in [noconf.ml](https://github.com/mattam82/Coq-Equations/blob/921ef5f5a7351d1dcef386c0fad4cfc481cf37c6/src/noconf.ml#L67) via [equations_common.ml])(https://github.com/mattam82/Coq-Equations/blob/921ef5f5a7351d1dcef386c0fad4cfc481cf37c6/src/equations_common.ml#L1049)
- once redefined with following code, doing `lift` and `subst` (in [principles.ml](https://github.com/mattam82/Coq-Equations/blob/921ef5f5a7351d1dcef386c0fad4cfc481cf37c6/src/principles.ml#L1042)):
```ocaml
let smash_rel_context sigma ctx =    
  let open Context.Rel.Declaration in 
  List.fold_right                             
    (fun decl (subst, pats, ctx') ->
       match get_value decl with    
       | Some b ->                           
         let b' = substl subst b in     
         (b' :: subst, List.map (lift_pat 1) pats, ctx')  
       | None -> (mkRel 1 :: List.map (lift 1) subst, 
                  PRel 1 :: List.map (lift_pat 1) pats,      
                  map_constr (Vars.substl subst) decl :: ctx'))  
    ctx ([], [], [])
```
A compromise could be to use this latter version but propagating an additional boolean telling to do no lift nor substitution until we've actually seen a `let`. What do you think?

Reply:
Well, I finally used an `Esubst`-based function which should be linear in the number of declarations and, if ever there is a let, linear also in the size of terms.